### PR TITLE
Install Danger locally for better performances on CI

### PR DESCRIPTION
**Pull request by mokagio**
https://github.com/wordpress-mobile/WordPress-iOS/pull/13423 added two [GitHub workflows](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions) to run static checks on PRs via [Danger](https://danger.systems/js/).

To keep the setup simple, I decided to use the [Danger GitHub Action](https://github.com/marketplace/actions/danger-js-action), but the time it takes to run on CI is too much, more than a minute.

This diff switches the setup to a local installation of Danger, which runs on CI in less than 30 seconds –after the first run, when the cache has been build.

Before | After | After, with cache
--- | --- | ---
Screen Shot 2020-02-17 at 11 17 21 | Screen Shot 2020-02-17 at 11 15 26 | Screen Shot 2020-02-17 at 11 18 37

This approach adds a bit of overhead on the repo: another dependency manager. Devs don't need to interact with it though, only CI does. I think it's an acceptable tradeoff to deliver feedback on the PR faster.

### How to test

Because this Danger setup runs on [pull request events](https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request), you're better off testing this changes in a new PR. It will keep this PR cleaner. Feel free to use https://github.com/mokagio/WordPress-iOS/pull/9 as a base for it, the code is the same, just less tidy.

The test strategy should be the same as #13442:

- Add and remove labels to verify Danger leaves comments when the PR has no label.
- Change the codebase in a way that will trigger a warning or failure in the consistency checks.

### CODE REVIEW

File: .github/workflows/danger-pr-labels.yml [hunk 1]
```diff
@@ -10,9 +10,32 @@ jobs:
...
         with:
-          args: "--dangerfile Automattic/peril-settings/org/pr/label.ts"
+          path: node_modules
+          key: ${{ runner.os }}-node_modules-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: ${{ runner.os }}-node_modules
+
+      - name: Yarn Install
+        if: steps.cache-node-modules.outputs.cache-hit != 'true'
+        # frozen-lockfile will make the build fail if the lockfile is not there
+        run: yarn install --frozen-lockfile
```
Comment (mokagio) [inline .github/workflows/danger-pr-labels.yml]: Only the output of `yarn install` is cached. I thought about making the `npm` command to install Yarn be a local one, to cache that output as well, but GitHub has already optimized `npm` runs and put in place their own caching for package used as much as Yarn.

Screen Shot 2020-02-17 at 11 03 38

File: .github/workflows/danger-prs.yml [hunk 2]
```diff
@@ -2,17 +2,40 @@ name: iOS Consistency Checks
 on:
   pull_request:
     # The restrictions on the types of pull_request event that can trigger this
-    # workflow is intentional. We only want to run these sanity checks when the
```
Comment (mokagio) [inline .github/workflows/danger-prs.yml]: More inclusive language. Thanks again @jkmassel for pointing this out.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "devDependencies": {
+    "danger": "^9.2.10",
+    "typescript": "^3.7.5"
```
Comment (mokagio) [inline package.json]: Notice that, on top of Danger, we have TypeScript as well. That's because [our `Dangerfile`s are written in TypeScript](https://github.com/Automattic/peril-settings/tree/5d22258ca46c3649f2cb863cfac2d16f12c4cba0), so Danger needs the TypeScript transpiler to be in the `node_modules/` folder in order to process them.

File: .github/workflows/danger-prs.yml (hunk 2 above)
Comment (jkmassel) [reply .github/workflows/danger-prs.yml]: Nice catch! 👍

File: .github/workflows/danger-pr-labels.yml (hunk 1 above)
Comment (jkmassel) [reply .github/workflows/danger-pr-labels.yml]: Given the fact that this is 1 second of the CI time, this seems like a sensible decision!

File: .github/workflows/danger-pr-labels.yml [hunk 4]
```diff
@@ -10,9 +10,32 @@ jobs:
...
+
+      - name: Validate PR Labels
```
Comment (jkmassel) [inline .github/workflows/danger-pr-labels.yml]: This differs from the job name which is "Validated PR Labels". IMHO saying "Validate" is better as it's a present-tense verb. Can we use it everywhere? :)

**Comment by jkmassel** [comment]
Note for future – this PR was tested using https://github.com/wordpress-mobile/WordPress-iOS/pull/13494.

File: .github/workflows/danger-pr-labels.yml (hunk 4 above)
Comment (mokagio) [reply .github/workflows/danger-pr-labels.yml]: Great suggestion. I do like the present-tense verb approach ❤️ . I'm onto it. 👍 .

Comment (mokagio) [reply .github/workflows/danger-pr-labels.yml]: Oh :o. Just seen that past-tense "Validate_d_" in the code. Not intentional, just a typo. Good catch 😄

**Comment by mokagio** [comment]
> One potential area for improvement for you to consider (and sorry if you've addressed this elsewhere) – while it's nice to run these in parallel, a lot of the overhead work (checking out the repo, setting up node, etc) is repeated for each job. **Running all the checks in the same job** would slightly increase run time, but given the degree to which you've optimized things, would that be worth it? Part of the benefit here would be grouping feedback from all of the actions into the same GitHub comment (in the same way that Peril does now). WDYT about this? 
(_emphasis mine_)

Thanks @jkmassel, that's a great suggestion. 

**TL;DR:** 👉 I like the idea and will try it out in a dedicated PR.

---

I ended up with this setup because I let the events that might trigger a workflow build drive which Danger file to run.

We run the check for the labels only when the labels can change:

https://github.com/wordpress-mobile/WordPress-iOS/blob/ca1f915437ffe8ffe8a7fe250b2bebfd07f621e2/.github/workflows/danger-pr-labels.yml#L2-L7

In the same fashion, we run the check on the code only when the code can change, not when the labels change:

https://github.com/wordpress-mobile/WordPress-iOS/blob/ca1f915437ffe8ffe8a7fe250b2bebfd07f621e2/.github/workflows/danger-prs.yml#L2-L7

I think this is less wasteful, at least from a certain point of view. Here's a PR life-cycle example

Event | Danger runs
--- | ---
open | labels validation, consistency checks
add label | labels validation
push new commits | consistency checks
push new commits | consistency checks
reviewed with change request | none
push new commits to address feedback | consistency checks
reviewed with approval | none
merge | none

In such a case, I think we'd save time by not running the label validation on every change to the PR.

But ☝️ the point you raise is totally legit. Actually, **given that the time it takes to run Danger is very little compared to the rest of the workflow setup and that those multiple comments are very noisy, I think the trade-off of running Danger checks that will produce the same result because nothing has changed is worth ends up being worth it**. If anything, it's definitely less usage on CI, not that we pay for it, but that's not an excuse to waste resources 🔋 🌏 ♻️ .

There's only one little thing I'm not sure about. I don't know if I can tell Danger to run more than one `Dangerfile` with the `--dangerfile` option. I'm going to look into it, then submit a dedicated PR.

**Comment by mokagio** [comment]
@jkmassel yeah. I don't have merge rights though, so you'll have to merge this for me :pray:
